### PR TITLE
Add support for reading NC_STRING attribute values

### PR DESCRIPTION
**Pull request by samoylovfp**
Hi! I was getting error `type 12 is not known` when trying to read NC_STRING typed attribute values.
This fixed that. Please mention if I missed anything.

### CODE REVIEW

File: src/attribute.rs [hunk 1]
```diff
@@ -408,7 +410,33 @@ impl<'a> Attribute<'a> {
...
+                        if str_ptr.is_null() {
+                            break;
+                        }
+                        let c_str = CStr::from_ptr(*str_ptr);
+                        result.push(c_str.to_string_lossy().to_string());
+                    }
+                    nc_free_string(attlen, buf.as_mut_ptr());
+                }
+                Ok(AttrValue::Str(
+                    result.join("\n")
```
Comment (samoylovfp) [inline src/attribute.rs:437]: This is an assumption. Do we want to have a separate `AttrValue` variant?

Comment (magnusuMET) [reply src/attribute.rs:437]: We should add `Strs(Vec<String>)` here, since a string attribute can consist of multiple `nc_string`

Comment (samoylovfp) [reply src/attribute.rs]: Implemented. Please, review.

File: src/attribute.rs [hunk 2]
```diff
@@ -815,6 +860,12 @@ impl From<String> for AttrValue {
     }
 }
 
+impl From<Vec<String>> for AttrValue {
```
Comment (magnusuMET) [inline src/attribute.rs:863]: Note for future: This can be made more general using a different type, e.g. `AsRef<[&str]>`

File: src/attribute.rs [hunk 3]
```diff
@@ -654,6 +681,24 @@ impl<'a> Attribute<'a> {
                         x.as_ptr() as *const _,
                     )
                 }),
+                AttrValue::Strs(ref x) => {
+                    let mut cstrings: Vec<CString> = Vec::with_capacity(x.len());
```
Comment (lnicola) [inline src/attribute.rs:685]: This is okay, but I think you can also use iterators here, see https://github.com/georust/gdal/pull/239/files#diff-c9786f4e33a40ebd8381383ee08e3f97174ff918b73664ba1b53019e10c76e5cR26-R40 for an example.

File: src/attribute.rs [hunk 4]
```diff
@@ -409,6 +411,30 @@ impl<'a> Attribute<'a> {
...
+                            self.ncid,
+                            self.varid,
+                            self.name.as_ptr() as *const _,
+                            buf.as_mut_ptr() as *mut *mut _,
+                        )
+                    }))?;
+
+                    for str_ptr in &buf {
+                        if str_ptr.is_null() {
+                            break;
```
Comment (magnusuMET) [inline src/attribute.rs:429]: I would rather push an emptry string to `result` in this case than skipping it

File: src/attribute.rs (hunk 3 above)
Comment (samoylovfp) [reply src/attribute.rs:685]: Ah always forget that I can collect into a result. I'll amend the commit, this looks much better IMO

File: src/attribute.rs (hunk 4 above)
Comment (magnusuMET) [reply src/attribute.rs:429]: And this can also be made into an iterator

**Review by magnusuMET — COMMENTED** [review]
Looks great, just a couple of nitpicks and this should be good to go!

File: src/attribute.rs (hunk 3 above)
Comment (samoylovfp) [reply src/attribute.rs:685]: Didn't amend, added a separate commit

File: src/attribute.rs [hunk 5]
```diff
@@ -409,6 +411,34 @@ impl<'a> Attribute<'a> {
...
+                        .map(|cstr_pointer| {
+                            if cstr_pointer.is_null() {
+                                String::new()
+                            } else {
+                                CStr::from_ptr(*cstr_pointer).to_string_lossy().to_string()
+                            }
+                        })
+                        .collect();
+
+                    nc_free_string(attlen, buf.as_mut_ptr());
```
Comment (magnusuMET) [inline src/attribute.rs:438]: You must `super::with_lock` this

Comment (samoylovfp) [reply src/attribute.rs:438]: Does it make sense to hold the lock while copying the data over to rust or should we take it again here?

Comment (magnusuMET) [reply src/attribute.rs:438]: No, it is only needed when calling functions from `netcdf-c`, we should have sole ownership of the data following `nc_get_att_string`

Comment (magnusuMET) [reply src/attribute.rs:438]: We can also ignore the error handling in this case: We don't care about a potential memory leak

Comment (samoylovfp) [reply src/attribute.rs:438]: Looks like performing several fallible operations in one `with_lock` invocation is difficult, thinking about adding another helper

Comment (magnusuMET) [reply src/attribute.rs:438]: You only have to wrap `nc_free_string`, everything else is fine here. The lock is just there to synchronise `nc_*` calls

Comment (samoylovfp) [reply src/attribute.rs]: Does this make sense?
https://github.com/georust/netcdf/pull/85/commits/bc7b89d0ccd73589b75b2ecf43242485f244cd99

Comment (samoylovfp) [reply src/attribute.rs]: Ah ok, changing to that

Comment (samoylovfp) [reply src/attribute.rs]: Looks like it always returns 0
https://www.unidata.ucar.edu/software/netcdf/docs/dvar_8c_source.html#l01206

Comment (magnusuMET) [reply src/attribute.rs]: No worries with this then :)

File: src/attribute.rs (hunk 2 above)
Comment (magnusuMET) [reply src/attribute.rs:868]: Running into problems with `AsRef` (upstream may add a new impl of trait), the following two impls should along with the two you've written cover most of the usecases:
```rust
impl From<&[&str]> for AttrValue {
    fn from(x: &[&str]) -> Self {
        Self::Strs(x.iter().map(|&s| String::from(s)).collect())
    }
}

impl From<Vec<&str>> for AttrValue {
    fn from(x: Vec<&str>) -> Self {
        Self::from(x.as_slice())
    }
}
```

Comment (samoylovfp) [reply src/attribute.rs:867]: Added

**Review by magnusuMET — APPROVED** [review]

**Comment by magnusuMET** [comment]
Great, thanks for the work @samoylovfp! When CI is green I'll merge and release a new version if you'd like

**Comment by samoylovfp** [comment]
Yes I would like a new version. Thanks a lot for your time! :crab:

**Comment by magnusuMET** [comment]
@samoylovfp Version `0.7.0` has now been published

**Pull request merged**